src/com/facebook/buck/file/downloader/impl/OnDiskGradleDownloader.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2015-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain
+ * a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.facebook.buck.file.downloader.impl;
+
+import com.facebook.buck.core.config.BuckConfig;
+import com.facebook.buck.core.exceptions.HumanReadableException;
+import com.facebook.buck.file.downloader.Downloader;
+import com.google.common.base.Strings;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Optional;
+
+/** A {@link Downloader} that pulls content in gradle format from the local file system. */
+public class OnDiskGradleDownloader extends OnDiskDownloader {
+  private static final String GRADLE_HOME = "GRADLE_USER_HOME";
+
+  public OnDiskGradleDownloader(Path root) throws FileNotFoundException {
+    super(root);
+  }
+
+  @Override
+  public Path getTargetPath(URI uri) {
+    // Get the relative path that the maven item is located at.
+    String relativePath = MavenUrlDecoder.toLocalGradlePath(uri);
+
+    Path fullPath = getRoot().resolve(relativePath);
+
+    try {
+      // Gradle keeps artifacts in the below format
+      // <group>/<artifactId>/<version>/hash/<filename>
+      // List all the folders in the parent directory, iterate and
+      // append hash & filename and return if the file is present.
+
+      if (fullPath.getParent().toFile().exists()) {
+        Optional<Path> actualPath =
+            Files.list(fullPath.getParent())
```

Comment:
is it possible to use hash here to accomodate the case where gradle cache may have different versions of the same artifact present?

Reply:
gradle and buck seems to be using different hashing algorithm, so the hashes don't match